src/adam_core/dynamics/moid.py

```diff
@@ -145,13 +165,129 @@ def calculate_moid(
         lambda dt: calculate_moid_for_dt(primary_ellipse, secondary_ellipse, dt),
         bounds=bounds,
         method="bounded",
-        tol=1e-12,
+        tol=1e-14,
     )
     moid = result.fun
     moid_time = Timestamp.from_mjd(
-        [primary_ellipse.coordinates.time.mjd()[0].as_py() + result.x], scale="tdb"
+        [primary_ellipse.coordinates.time.mjd()[0].as_py() + result.x],
+        scale=primary_ellipse.coordinates.time.scale,
     )
     if result.status != 0:
         raise ValueError("MOID calculation did not converge.")
 
     return moid, moid_time
+
+
+def moid_worker(
+    idx_chunk: npt.NDArray[np.int64], orbits: Orbits, perturber: OriginCodes
+) -> PerturberMOIDs:
+    """
+    Calculate the MOID for a chunk of orbits with respect to a perturbing body.
+    """
+    orbits_chunk = orbits.take(idx_chunk)
+    states = get_perturber_state(
+        perturber,
+        orbits_chunk.coordinates.time,
+        frame=orbits_chunk.coordinates.frame,
+        origin=orbits_chunk.coordinates.origin[0].as_OriginCodes(),
+    )
+    moids = PerturberMOIDs.empty()
+    for orbit, state in zip(orbits_chunk, states):
+        moid, moid_time = calculate_moid(
+            orbit, Orbits.from_kwargs(orbit_id=[perturber.name], coordinates=state)
+        )
+
+        moids_i = PerturberMOIDs.from_kwargs(
+            orbit_id=orbit.orbit_id,
+            perturber=Origin.from_kwargs(code=[perturber.name]),
+            moid=[moid],
+            time=moid_time,
+        )
+
+        moids = qv.concatenate([moids, moids_i])
+
+    return moids
+
+
+moid_worker_ray = ray.remote(moid_worker)
+
+
+def calculate_perturber_moids(
+    orbits: Orbits,
+    perturber: Union[OriginCodes, List[OriginCodes]],
+    chunk_size: int = 100,
+    max_processes: int = 1,
+) -> PerturberMOIDs:
+    """
+    Calculate the minimum orbit intersection distance (MOID) for all orbits with respect
+    to the perturbing body or bodies.
+
+    Parameters
+    ----------
+    orbits : Orbits
+        The orbits to calculate the MOID for.
+    perturber : OriginCodes or List[OriginCodes]
+        The perturbing body or bodies to calculate the MOID for.
+    chunk_size : int, optional
+        The number of orbits to process in each chunk, by default 100.
+    max_processes : int, optional
+        The maximum number of processes to use, by default 1.
+
+    Returns
+    -------
+    PerturberMOIDs
+        A table containing the MOID and time for each orbit with respect to the perturbing body or bodies.
+    """
+    assert len(orbits.orbit_id.unique()) == len(orbits)
+    assert len(orbits.coordinates.origin.code.unique()) == 1
```

Comment:
Probably something obvious, but why this assert?

Reply:
If we wanted to accept arbitrary origins then we would need to batch by the orbits' unique origin as well (since we need to pull the perturber states to have the same origin). I suppose this MOID calculation really only makes sense from the heliocenter or barycenter at the moment.